### PR TITLE
MDEV-23175: my_timer_milliseconds ftime deprecated - clock_gettime re…

### PR DESCRIPTION
**Pull request by grooverdan**
…placement

Linux glibc has deprecated ftime resutlting in a compile error on Fedora-32.

Per manual clock_gettime is the suggested replacement. Because my_timer_milliseconds
is a relative time used by largely the perfomrance schema, CLOCK_MONOTONIC_COARSE
is used. This has been available since Linux-2.6.32.

The low overhead is shows in the unittest:

    $ unittest/mysys/my_rdtsc-t
    1..11
    # ----- Routine ---------------
    # myt.cycles.routine          :             5
    # myt.nanoseconds.routine     :            11
    # myt.microseconds.routine    :            13
    # myt.milliseconds.routine    :            18
    # myt.ticks.routine           :            17
    # ----- Frequency -------------
    # myt.cycles.frequency        :    3598238805
    # myt.nanoseconds.frequency   :    1000000000
    # myt.microseconds.frequency  :       1000000
    # myt.milliseconds.frequency  :          1000
    # myt.ticks.frequency         :           102
    # ----- Resolution ------------
    # myt.cycles.resolution       :             1
    # myt.nanoseconds.resolution  :             1
    # myt.microseconds.resolution :             1
    # myt.milliseconds.resolution :          1000
    # myt.ticks.resolution        :             1
    # ----- Overhead --------------
    # myt.cycles.overhead         :            32
    # myt.nanoseconds.overhead    :            64
    # myt.microseconds.overhead   :            58
    # myt.milliseconds.overhead   :             8
    # myt.ticks.overhead          :          1362
    ok 1 - my_timer_init() did not crash
    ok 2 - The cycle timer is strictly increasing
    ok 3 - The cycle timer is implemented
    ok 4 - The nanosecond timer is increasing
    ok 5 - The nanosecond timer is implemented
    ok 6 - The microsecond timer is increasing
    ok 7 - The microsecond timer is implemented
    ok 8 - The millisecond timer is increasing
    ok 9 - The millisecond timer is implemented
    ok 10 - The tick timer is increasing
    ok 11 - The tick timer is implemented

### CODE REVIEW

File: mysys/my_rdtsc.c [hunk 1]
```diff
@@ -175,7 +175,11 @@ ulonglong my_timer_microseconds(void)
 
 ulonglong my_timer_milliseconds(void)
 {
-#if defined(HAVE_SYS_TIMEB_H) && defined(HAVE_FTIME)
+#if defined(HAVE_CLOCK_GETTIME)
+  struct timespec tp;
+  clock_gettime(CLOCK_MONOTONIC_COARSE, &tp);
+  return (ulonglong)tp.tv_sec * 1000 + (ulonglong)tp.tv_nsec / 1000;
```
Comment (kevgs) [inline mysys/my_rdtsc.c:181]: Nanoseconds divided by 1000 became microseconds. I think you should divide by 1000000 here.

**Comment by dr-m** [comment]
A comment at the end of `mysys/my_rdtsc.c` that was added in the initial version in 2009 (commit 6e01be7552aa5c5fb854d30d5d8cb61c2729b292) says:
```
   ftime() -- A "man ftime" says: "This function is obsolete.
   Don't use it." On every platform that we tested, if ftime()
   was available, then so was gettimeofday(), and gettimeofday()
   overhead was always at least as good as ftime() overhead.
```
I see that you did not remove `HAVE_FTIME` altogether. Does AIX support an alternative by now?

**Comment by grooverdan** [comment]
Seems AIX has clock_gettime (https://www.ibm.com/support/knowledgecenter/ssw_aix_72/c_bostechref/clock_getres.html)

*Solaris needs a fallback to CLOCK_REALTIME  - http://www.polarhome.com/service/man/?qf=clock_gettime&tf=2&of=Solaris&sf=
https://www.unix.com/man-page/opensolaris/3c/clock_gettime/

FreeBSD - https://www.freebsd.org/cgi/man.cgi?query=clock_gettime need to look closer at its ticks concept.

**Comment by kevgs** [comment]
What a maintenance burden dealing with all these type of clocks!

I'd accept this patch as is because it fixes a real issue. I have such a compilation problem locally.

And in the near future we should consider using `std::chrono::steady_clock` everywhere. This will also fix a performance problem with calling the slowest `times()` function multiple times during a server start up. I can see this function in profiler!

**Comment by dr-m** [comment]
I agree that perfect is the enemy of good, and maybe we could accept this as is.

If @grooverdan wants to clean up it a little further and remove all traces of `ftime()`, I think that it might be acceptable in 10.5, even though 10.5.4 was a GA release. For 10.6 there should be no doubt.

**Comment by grooverdan** [comment]
I'll scope this down to a linux only use of `clock_gettime` for 10.5. As-is its going to break some platforms due to limited clockids everywhere.

For 10.6 a full removal of `ftime()` is a good goal either by `std::chrono::steady_clock` or indivdual clock_ids  (for which I'll do some micro testing on solaris/FreeBSD  to ensure there really is microseconds available on `clock_gettime`). I will extend the unit test regardless to ensure there is some precisions.

**Comment by ericherman** [comment]
I appreciate the pragmatic 10.5 vs 10.6 approach as this makes it reasonable to get "good" in now and also plan for "better" in the near future. Thank you.

**Comment by kevgs** [comment]
Thank you!

**Pull request merged**